pkg/services/ngalert/schedule/schedule.go

```diff
@@ -653,7 +635,38 @@ func (r *alertRuleRegistry) keyMap() map[models.AlertRuleKey]struct{} {
 type alertRuleInfo struct {
 	evalCh  chan *evalContext
 	stopCh  chan struct{}
-	version int64
+	mtx     sync.Mutex
+	stopped bool // indicates that the stopCh was closed and therefore the rule evaluation routine should exit
+}
+
+func newAlertRuleInfo() *alertRuleInfo {
+	return &alertRuleInfo{evalCh: make(chan *evalContext), stopCh: make(chan struct{}), stopped: false}
+}
+
+// Stop signals the rule evaluation routine to stop via stopCh and sets flag stopped to true. Does nothing If the loop is stopped
+func (a *alertRuleInfo) Stop() bool {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+	if a.stopped {
+		return false
+	}
+	close(a.stopCh)
+	a.stopped = true
+	return true
+}
+
+// Eval signals the rule evaluation routine to perform the evaluation of the rule. Does nothing if the loop is stopped
+func (a *alertRuleInfo) Eval(t time.Time, version int64) bool {
+	a.mtx.Lock()
```

Comment:
Here I don't think we need to acquire the mutex, we can just check if `a.stopCh` is closed. `a.stopped` still needs to be set to true to make sure we don't close a closed channel on successive calls to `Stop()`.

```suggestion
	select {
	    case <-a.stopCh:
	        return false
	    default:
	}
```